scripts/igmp_query_injector.py

```diff
@@ -1,41 +1,85 @@
 #!/usr/bin/env python
-import Queue
 import argparse
-import os
-import socket
-import struct
-import subprocess
 import threading
 from abc import ABCMeta, abstractmethod
 import logging
 import signal
-import resource
-import binascii
+from scapy.all import Ether, IP, sendp
 from xcp import logger as log
-import time
 import sys
 import re
 from xen.lowlevel.xs import xs
 
 
+#################################
+# Code for build IGMP layer
+from scapy.packet import *
+from scapy.fields import *
+from scapy.layers.inet import *
+
+class IGMP(Packet):
+  """IGMP Message Class for v1 and v2.
+
+    Parameters:
+      type    IGMP type field, 0x11, 0x12, 0x16 or 0x17
+      mrtime  Maximum Response time (zero for v1)
+      gaddr   Multicast Group Address 224.x.x.x/4
+      
+See RFC2236, Section 2. Introduction for definitions of proper 
+IGMPv2 message format   http://www.faqs.org/rfcs/rfc2236.html
+
+  """
+  name = "IGMP"
+
+  igmptypes = { 0x11 : "Group Membership Query",
+                0x12 : "Version 1 - Membership Report",
+                0x16 : "Version 2 - Membership Report",
+                0x17 : "Leave Group"}
+
+  fields_desc = [ ByteEnumField("type", 0x11, igmptypes),
+                      ByteField("mrtime",20),
+                    XShortField("chksum", None),
+                        IPField("gaddr", "0.0.0.0")]
+
+  def post_build(self, p, pay):
+    """Called implicitly before a packet is sent to compute and place IGMP checksum.
+
+    Parameters:
+      self    The instantiation of an IGMP class
+      p       The IGMP message in hex in network byte order
+      pay     Additional payload for the IGMP message
+    """
+    p += pay
+    if self.chksum is None:
+      ck = checksum(p)
+      p = p[:2]+chr(ck>>8)+chr(ck&0xff)+p[4:]
+    return p
+
+bind_layers(IP, IGMP, frag=0, proto=2)
```

Comment:
I am thinking if it is better to group the global calls somewhere more visible, such as before the main for readability purpose.

Reply:
I'd like to keep it because this block is copy & crop from `scapy.contrib` package.